### ISSUE TITLE
Errors with declarations when compiling from source

### ISSUE THREAD

**Issue by gde061** [evgeni/tp_smapi#26]
I have an older ThinkPad T43 running Fedora 12 (Kernel 2.6.32.26-175.fc12.i686).  I am trying to get tp_smapi set up on this machine as I just replaced the battery and want to set charging thresholds. 

Given the age of the machine, my yum repositories for this version of Fedora don't seem to have a package I can install without compiling from source.  If there is a repository version that might work with my ancient machine, please let me know what repository to add, as that's the way I'm most familiar with installing stuff. 

I have followed the instructions at http://www.thinkwiki.org/wiki/Tp_smapi#Installation, namely the first step to install kernel headers and makefile with 
`yum install kernel-devel-$(uname -r)`

That appeared to work, but when I tried to compile tp_smapi it fails with lots of declaration errors. See below for specific errors from `make load`.  I'm not an expert at dependencies, so I would really appreciate it if someone could help me figure out what I'm missing.  I suspect it is some other package I overlooked which has the missing functions?  

Thanks in advance!

> if lsmod | grep -q '^hdaps '; then rmmod hdaps; fi
> if lsmod | grep -q '^tp_smapi '; then rmmod tp_smapi; fi
> if lsmod | grep -q '^tp_base '; then rmmod tp_base; fi
> make -C /lib/modules/2.6.32.26-175.fc12.i686/build M=/home/snowyd/tp_smapi-0.25 modules
> make[1]: Entering directory `/usr/src/kernels/2.6.32.26-175.fc12.i686'
>   CC [M]  /home/snowyd/tp_smapi-0.25/tp_base.o
> /home/snowyd/tp_smapi-0.25/tp_base.c:90: warning: type defaults to ‘int’ in declaration of DECLARE_MUTEX’

> /home/snowyd/tp_smapi-0.25/tp_base.c:90: warning: parameter names (without types) in function declaration
> /home/snowyd/tp_smapi-0.25/tp_base.c: In function ‘tp_controller_lock’:
> /home/snowyd/tp_smapi-0.25/tp_base.c:99: error: implicit declaration of function ‘down_interruptible’
> /home/snowyd/tp_smapi-0.25/tp_base.c:99: error: ‘tp_controller_mutex’ undeclared (first use in this function)
> /home/snowyd/tp_smapi-0.25/tp_base.c:99: error: (Each undeclared identifier is reported only once
> /home/snowyd/tp_smapi-0.25/tp_base.c:99: error: for each function it appears in.)
> /home/snowyd/tp_smapi-0.25/tp_base.c: In function ‘tp_controller_try_lock’:
> /home/snowyd/tp_smapi-0.25/tp_base.c:113: error: implicit declaration of function ‘down_trylock’
> /home/snowyd/tp_smapi-0.25/tp_base.c:113: error: ‘tp_controller_mutex’ undeclared (first use in this function)
> /home/snowyd/tp_smapi-0.25/tp_base.c: In function ‘tp_controller_unlock’:
> /home/snowyd/tp_smapi-0.25/tp_base.c:123: error: implicit declaration of function ‘up’
> /home/snowyd/tp_smapi-0.25/tp_base.c:123: error: ‘tp_controller_mutex’ undeclared (first use in this function)
> /home/snowyd/tp_smapi-0.25/tp_base.c: In function ‘dmi_find_substring’:
> /home/snowyd/tp_smapi-0.25/tp_base.c:386: warning: assignment discards qualifiers from pointer target type
> make[2]: *** [/home/snowyd/tp_smapi-0.25/tp_base.o] Error 1
> make[1]: *** [_module_/home/snowyd/tp_smapi-0.25] Error 2
> make[1]: Leaving directory `/usr/src/kernels/2.6.32.26-175.fc12.i686'
> make: *** [modules] Error 2

>

**Comment by gde061** [comment]
To answer the first part of my question about a repository version, I stumbled onto this thread [(http://forum.thinkpads.com/viewtopic.php?t=83009)] about an AKMOD version... don't know what that is but the discussion said it worked on Fedora 12.  

Following the link to [https://izhar.fedorapeople.org/tp_smapi-kmod/], I tried first the PAE version, which would not install because of my kernel version.  But the first one did install after getting a bunch of other dependencies.  Following the guide in the first thread on how to start the think, it seemed to accept the echo commands.  Now I just have to watch the battery charge levels to see if it's really working.  

I'd still love to get an answer about the declarations though, as the main reason I have the machine is to learn more about linux.   Thanks.
